class Solution {
    public int longestValidParentheses(String s) {
        int n = s.length();

        int open = 0;
        int close = 0;

        int result = 0;

        for(int i=0; i<n; i++) {
            char ch = s.charAt(i);

            if(ch == '(') {
                open++;

            } else {
                close++;
            }

            if(open == close) {
                result = Math.max(result, open+close);

            } else if(close > open) {
                open = close = 0;

            } 
            
        }

        open = 0;
        close = 0;

        for(int i = n-1; i >=0; i--) {

            char ch = s.charAt(i);

            if(ch == '(') {
                open++;
            }
            else {
                close++;
            }

            if(open == close) {
                result = Math.max(result, open+close);

            } else if( open > close) {
                open = close = 0;
            }
        }

        return result;

  
    }
}